File: day03/js/solution.js
import fs from "fs";

const input = fs
  .readFileSync("../input.txt", "utf-8")
  .trim()
  .split("\n")
  .map((line) => line.trim());

function maxJoltage(line) {
  const digits = line.split("").map(Number);
  let best = 0;

  for (let i = 0; i < digits.length - 1; i++) {
    let maxRight = 0;

    for (let j = i + 1; j < digits.length; j++) {
      maxRight = Math.max(maxRight, digits[j]);
    }

    best = Math.max(best, digits[i] * 10 + maxRight);
  }

  return best;
}

function solve(lines) {
  return lines.reduce((sum, line) => sum + maxJoltage(line), 0);
}

console.log("Result:", solve(input));

File: day03/py/solution.py
from pathlib import Path

lines = Path("../input.txt").read_text().strip().splitlines()

def max_joltage(line):
    digits = list(map(int, line))
    best = 0

    for i in range(len(digits) - 1):
        max_right = max(digits[i+1:])
        best = max(best, digits[i] * 10 + max_right)

    return best

def solve(lines):
    return sum(max_joltage(line) for line in lines)

print("Result:", solve(lines))
